src/Microsoft.AspNetCore.SystemWebAdapters.CoreServices/HttpRuntimeFactory.cs

```diff
@@ -2,42 +2,60 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 
 using System;
+using System.Web.Caching;
 using Microsoft.AspNetCore.SystemWebAdapters;
 
 namespace Microsoft.Extensions.DependencyInjection;
 
 internal static class HttpRuntimeFactory
 {
-    public static IHttpRuntime Create()
+    public static IHttpRuntime Create(IServiceProvider serviceProvider)
     {
         if (NativeMethods.IsAspNetCoreModuleLoaded())
         {
             var config = NativeMethods.HttpGetApplicationProperties();
-
-            return new IISHttpRuntime(config);
+            return new IISHttpRuntime(config, serviceProvider);
         }
 
-        return new DefaultHttpRuntime();
+        return new DefaultHttpRuntime(serviceProvider);
+    }
+
+    internal abstract class BaseHttpRuntime : IHttpRuntime
+    {
+        private readonly IServiceProvider serviceProvider;
+        private Cache? cache;
+
+        protected BaseHttpRuntime(IServiceProvider serviceProvider)
+        {
+            this.serviceProvider = serviceProvider;
+        }
+        public abstract string AppDomainAppVirtualPath { get; }
```

Comment:
This type itself doesn't need to implement IHttpRuntime, just the stuff that is shared and requires the service provider. The IHttpRuntime can be just added to the DefaultHttpRuntime/IIS